Implement RPC methods for configuring the GBTX's

lpetre-ulb opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Addresses #72 

This PR adds a new RPC module which implements 6 new methods related to the GBTX configuration :

* `writeGBTConfig` updates the configuration of one chosen GBTX.
* `writeGBTConfigMultiLink`, multi-link version of the previous method.
* `writeGBTPhase` updates the RX phases of one OH.
* `writeGBTPhaseMultiLink`, multi-link version of the previous method.
* `scanGBTPhase` scans the phases on one OH.
* `scanGBTPhaseMultiLink`, multi-link version of the previous method.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue(s) here. -->

This PR fixes issue #72.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

These tests were performed to see if the functionalities implemented in the `gbt.py` were still working with the RPC modules and so the new CTP7 releases :

* Writing the GBTX configurations. It has been tested for different OH, multiple GBTX's at a time.
* Setting the GBTX RX phases for 1 VFAT at a time.
* Setting the GBTX RX phases for multiples VFAT's at a time.
* Scanning the GBTX RX phases for one OH.

More details are available in the `reg_utils` PR : https://github.com/cms-gem-daq-project/reg_utils/pull/42

The environment was the following :
* CTP7 : eagle63
* CTP7 firmware : 3.7.0
* CTP7 build-id : CTP7-GENERIC-20180529T153916-0500-4935611
* `xHAL` commit https://github.com/cms-gem-daq-project/xhal/pull/99/commits/59d788dea1cf0dfb1888c39eb9744f3e905abdcb
* `cmsgemos` commit https://github.com/cms-gem-daq-project/cmsgemos/pull/226/commits/66639b6672b2a9fdcdec2a02c67b93e9904a759f
* The new `gbt.py` used for these tests comes from commit https://github.com/cms-gem-daq-project/reg_utils/pull/42/commits/c4ad5f3afe2ba3236538c1336c37d081f015af68

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- Template thanks to https://www.talater.com/open-source-templates/#/page/99 -->

mexanick commented:
Actually, the modules themselves provide a sort of encapsulation as in order to perform a call the the caller should load an appropriate module and the call its method. I agree with @jsturdy that functional module separation should take place here for the better code organization.

lpetre-ulb commented:
Ok, thank you for the suggestion, I've moved the methods to a new RPC module and new files. I've also added all the "multi-link" variants (see the updated first post).

bdorney commented:
What's the status of this PR? This is a mission critical action and I was hoping that we would have made more progress on it.

lpetre-ulb commented:
I've performed the tests yesterday afternoon. Writing the configuration, setting and scanning the phases work as expected. I'm committing last changes before review. I'm also porting the `gbt.py` tool in `reg_utils` for using the new RPC methods.

lpetre-ulb commented:
I've updated the first post to keep it up-to-date with the status of the PR. The only missing part is the new `gbt.py` tool of which its PR should be ready before the end of the day.

lpetre-ulb commented:
With the last [PR for the `gbt.py` tool](https://github.com/cms-gem-daq-project/reg_utils/pull/42), the PR's are made in all four repositories. The GBT tool functionalities are working from the DAQ machine with the version 3.7.0 of the CTP7 firmware.

bdorney commented on include/gbt.h:
```diff
@@ -0,0 +1,90 @@
+std::vector<uint32_t> scanGBTPhaseLocal(localArgs *la, const uint32_t ohN, const uint8_t phaseMin, const uint8_t phaseMax, const uint8_t phaseStep);
+
+/*! \fn void scanGBTPhase(const RPCMsg *request, RPCMsg *response)
+ *  \brief Scan the GBT phases of one optical link
```
What is `optical link` here?  e.g. does this only scan `GBT2` of `OHX`?  Or does it scan for *all* `GBT`'s on `OHX`?  Please update documentation to improve clarity.

lpetre-ulb replied:
It scans _all_ the `GBT`'s on `ohN`. I used `optical link` because it was used somewhere else where it seems to means the OH index. I'll clarify that point.

bdorney commented on include/gbt.h:
```diff
@@ -0,0 +1,90 @@
+void writeGBTPhaseLocal(localArgs *la, const uint32_t ohN, const uint32_t vfatN, const uint8_t phase);
+
+/*! \fn void writeGBTPhase(const RPCMsg *request, RPCMsg *response)
+ *  \brief Write the configuration of one GBTX
```
Does this write the *full* configuration? e.g. all e-link phases on this `GBTx`? Or does it write just a single phase? From the functional arguments I assume the single phase case.  Please update the documentation to resolve this ambiguity.

lpetre-ulb replied:
It write just a single phase. I forgot to update the `\brief line`...

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    static constexpr std::array<std::array<uint16_t, 3>, 10> elinkToRegisters{
+        { {69, 73, 77}, {67, 71, 75}, {93, 97, 101}, {91, 95, 99}, {117, 121, 125}, {115, 119, 123}, {141, 145, 149}, {139, 143, 147}, {165, 169, 173}, {163, 167, 171} }
+    };
+}
```
These are specific to GE1/1 but we will have soon (Q1 2019) GE2/1 OH's that will have different hardware link assignments.  How do we intend to handle this?

One approach would be to use different namespaces for the different hardware.  Or a different naming convention.  I think this should be finalized *before* this PR is merged.

Beyond the scope of this PR I would like to see a header file that has all hard-coded hardware specific parameters in a single location; I suspect some refactoring would be needed to achieve this.

bdorney replied:
@jsturdy @mexanick

lpetre-ulb replied:
What could be done is to use `#ifdef` when the values are different between GE1/1 and GE2/1. It is then possible to chose at compile time if modules are for GE1/1 or GE2/1 (with the `-D` option of the compiler). In fact, this solution does not requires to centralize all the constants in a single location. It might however be cleaner, particularly if the constants are sorted within namespaces.

bdorney replied:
For this PR don't worry about it because it's outside the scope.  I just continue to raise this issue.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    };
+}
+
+void writeGBTRegLocal(localArgs *la, const uint32_t ohN, const uint32_t gbtN, const uint16_t address, const uint8_t value){
```
In the header file this is the last function you declare but in this file it's the first definition you provide.  Perhaps for maintainability this should follow the order in which it's declared.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    LOGGER->log_message(LogManager::INFO, stdsprintf("Writing the configuration of OH #%u - GBTX #%u.", ohN, gbtN));
+
+    for (size_t address = 0; address < config.size(); address++){
+        writeGBTRegLocal(la, ohN, gbtN, static_cast<uint16_t>(address), config[address]);
```
You're assuming that `config` matches the expected size and order (e.g. index 0 is address 0) without:

1. actually checking that the size meets expectation, or
2. including a statement in the function documentation stating how the elements of `config` should be ordered and what it's size is expected to be (e.g. less than 366).

lpetre-ulb replied:
Yes, right, it should be in documentation. The size is already checked in the non-local versions and using it with a incorrect size vector will only not work (but silently).

bdorney replied:
Right, but the local version could be called by some other function so it too should work as a standalone object with proper handling.

lpetre-ulb replied:
Ok, I'll do as you suggest.

But that raise a question for me. Where do you expect the checks to be done ? For example, in the `writeGBTRegLocal` function, I should check that `ohN` is inferior to the number of OH's supported by the CTP7. However, the performances will suffer from a systematic access to the `GEM_AMC.GEM_SYSTEM.CONFIG.NUM_OF_OH` register. This is the reason why I removed some checks before committing.

bdorney replied:
Yes it makes sense to have a register check only when necessary since this causes a transaction on the bus which is O(us) but the CTP7's Zynq processor is 667 MHz [[1]](http://iopscience.iop.org/article/10.1088/1748-0221/11/02/C02011/pdf) so checks that do not require register access are "cheap."

Does this clarify things?

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    if (request->get_binarydata_size("config") != 366){
+        LOGGER->log_message(LogManager::ERROR, "The configuration has not the correct size. It must be 366 8-bits registers long.");
+        response->set_string("error", "The configuration has not the correct size. It must be 366 8-bits registers long.");
+        return;
```
This information is *not* contained in the function documentation in the header file.  This should be addressed.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+        LOGGER->log_message(LogManager::ERROR, stdsprintf("The provided number of gbtMask's does not fullfill the expectations (%u).", ohUnmaskedNumber));
+        response->set_string("error", stdsprintf("The provided number of gbtMask's does not fullfill the expectations (%u).", ohUnmaskedNumber));
+        return;
+    }
```
If you look at the structure of all functions the rpc request and response have keys that match the local callable functions; and in that case their documentation is provided by the local callable function.  But here you're using the keys `ohMask` and `gbtMasks` that are neither documented in this function documentation nor used by the local callable function.

Consider the `\details` field in `doxygen` to provide this information.  A developer/expert user should be able to understand what is needed for successful execution from looking at the function API and right now this is not possible.

lpetre-ulb replied:
Ok, I will add these informations.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+        LOGGER->log_message(LogManager::ERROR, stdsprintf("The configuration has not the correct size. This method expects %i 8-bits registers.", expectedConfigSize));
+        response->set_string("error", stdsprintf("The configuration has not the correct size. This method expects %i 8-bits registers.", expectedConfigSize));
+        return;
+    }
```
Not included in function documentation.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    // gbtMasks key
+    if (request->get_word_array_size("gbtMasks") != ohUnmaskedNumber){
+        LOGGER->log_message(LogManager::ERROR, stdsprintf("The provided number of gbtMask's does not fullfill the expectations (%u).", ohUnmaskedNumber));
+        response->set_string("error", stdsprintf("The provided number of gbtMask's does not fullfill the expectations (%u).", ohUnmaskedNumber));
```
add the provided `gbtMasks` in these error statements.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+            }
+            ohConfigIdx++;
+        }
+    }
```
It seems like you're relying on someone to provide both an `ohMask` (12 bit number) and then a vector that contains 3-bit numbers, coupled with a check on the number of unmasked bits in the `ohMask`.  With this information you're looping over using this to define which GBT's are programmed.  But this is overly complicated in my view.  There's a better way.

All you need is the `gbtMasks` key.  Force the `i^th` element of `gbtMasks` to correspond to the `i^th` optohybrid.  Then a value of `0x0` implies the OH is masked; any nonzero value then implies the OH is not masked and provides the GBT's to be considered.

This requires minimal modification to the loop in Lines 134-147 while simplifying the rpc input.

lpetre-ulb replied:
Yes, that's a better possibility, even if the keys will loose in uniformity. However, I would still need to keep track of the number of OH or GBT's which have been configured. Except, it the user writes padding in the `config` blob at the place of masked OH's/GBT's, but the data will be less compressed.

bdorney replied:
I’m not sure I understand this issue. Could you elaborate?

lpetre-ulb replied:
My idea for the `writeGBTConfigMultiLink` function was to be able to use a different GBTX configuration file for each GBT to program, in a single RPC call. It allows to easily mix short and long chambers. These configurations must be concatenated in order to fit in key system, but there are multiple packing possibilities :
1. Concatenate all configurations. If the corresponding GBT is masked, you can add 366 zeroes, but the data is not compressed in any way.
2. Concatenate the 3 configurations only for the unmasked OH's. (This is what I did by imitation with the single link function call.)
3. Concatenate only the configurations of the unmasked GBT's. (This is what I plan to do.)

The the last two possibilities require to keep track of the position in the packed configuration data. But there are ways to keep the tracking clean.

bdorney replied:
Okay; I look forward to seeing the implementation.  But I think you got my primary concern: namely we don't need to provide two inputs that could potentially contradict.  This can be done just with `gbtMasks` and the `ohMasks` input is not needed.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    return;
+} //End writeGBTPhaseLocal
+
+void writeGBTPhase(const RPCMsg *request, RPCMsg *response){
```
How you have made this function is more like `writeAllGBTPhases(...)` and while this *is* potentially useful it was not what was requested in #72.  We are looking for a function that takes a single VFAT position and a phase value and sets that phase for the corresponding elink.

Keep this function, but rename it to  `writeAllGBTPhases(...)`, update the documentation to reflect that this is for *all* VFATs on an optohybrid and then make a new function `writeGBTPhase(...)` that is the remote callable method for writing just one VFAT worth of info (as requested in #72).

lpetre-ulb replied:
Ok, I'll add the function.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    return;
+} //End writeGBTPhase
+
+void writeGBTPhaseMultiLink(const RPCMsg *request, RPCMsg *response){
```
Rename to `writeAllGBTPhaseMultiLink` and update function documentation accordingly.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    const uint32_t ohMax = readReg(&la, "GEM_AMC.GEM_SYSTEM.CONFIG.NUM_OF_OH");
+
+    // vfatMask key
+    const std::bitset<32> vfatMask(request->get_word("vfatMask"));
```
Multi link functions should *not* take a `vfatMask` from the rpc message but determine the *optohybrid specific* `vfatMask` at runtime, see for example:

https://github.com/cms-gem-daq-project/ctp7_modules/blob/314cec21c7b565f898e461b487e1504907b78d6c/src/calibration_routines.cpp#L1282-L1284

bdorney replied:
(otherwise you *force* all OH's to use *the same* `vfatMask` which is not correct).

lpetre-ulb replied:
Indeed, forcing all the OH's to use the same `vfatMask` is maybe not the best solution. But using `getOHVFATMaskLocal` won't work. For example, after a phase scan, all the phases are left at value 15 which never works. I'll instead provide a array of `vfatMask` (like the gbtMasks).

bdorney replied:
Yes you’re right. On further reflection why do we need the VFAT mask at all here?  Can it be dropped? The only thing that seems to actually matter is the attempt to read CFG_RUN and this could be handled by catching in the bus error that would be raised by a failed read.  Or did I miss something?

bdorney replied:
I would propose dropping VFAT mask, accepting scanning over dead/empty VFATs, and catching bus errors.  I don’t think this would be a significant issue.

lpetre-ulb replied:
I don't understand exactly what you mean; this is the `writeAllGBTPhaseMultiLink` function. In the scan phase function, the scan indeed runs over the dead/empty VFAT's.

My idea behind an array of `vfatMask` is to be able to write some phases spread over all links in a single RPC call while leaving untouched the phases we do not want to change. Isn't it useful ?

bdorney replied:
Can you provide a use case in which you would *want* to do that?  This function in my mind is called at configure.  Then everything is configured.  In which case would you want to write only some phases?  I can't think of a case for that.

lpetre-ulb replied:
My use case was the correction of multiple VFAT phases in test stands while leaving the others at their default value. I could also argue that if the only use case is at configure, this `writeAllPhasesMultiLink` method is well performed with an on-the-fly modification of the configuration blob. But, OK, I'll remove the `vfatMasks`.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+        LOGGER->log_message(LogManager::ERROR, "There must be 24 VFAT phases by OH.");
+        response->set_string("error", "There must be 24 VFAT phases by OH.");
+        return;
+    }
```
Information not in function definition.

bdorney replied:
Moreover you have not described the expected ordering.

bdorney replied:
Change error message to:

`There must be 24 VFAT phases per OH`

e.g. `by OH` does not make sense.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+
+    std::vector<uint32_t> results;
+
+    for(uint8_t phase = phaseMin; phase <= phaseMax; phase += phaseStep){
```
At minimum you should check that `phaseMax` makes sense; e.g. an 8-bit number goes to 255 while the max phase that can be set is 15 (16? I don't remember if indexing starts at 1 or 0).

lpetre-ulb replied:
Of course, I missed that.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    return results;
+} //End scanGBTPhaseLocal
+
+void scanGBTPhase(const RPCMsg *request, RPCMsg *response){
```
I was expecting that this would *also* take a key that would determine the number of scans that would be performed as described [here](https://github.com/cms-gem-daq-project/ctp7_modules/issues/72#issuecomment-441643357).  The goal is to have a tool that makes `N` scans in a single RPC call (otherwise you lose a lot of time to the RPC message) and that this can be used to potentially determine the GBT phase settings.

We saw that between long & short boards the phase values differ and that even performing multiple scans on the *same* OH+GEB pair one after another give different results (edge effects).

lpetre-ulb replied:
I misunderstood that request; I understood it as the `phaseMin`, `phaseMax` and `phaseStep` parameters. I'll add a `N` scans parameters. As the number of phases is very small, does it worth keeping the `phaseMin`, `phaseMax` and `phaseStep` parameters ?

bdorney replied:
You’ve already coded them so if it’s easier to keep them why not; just assign default values to them in the header file, e.g. min 0, max 15, step size 1.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+        // Try to synchronise the VFAT's
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        writeReg(la, "GEM_AMC.GEM_SYSTEM.CTRL.LINK_RESET", 1);
+        std::this_thread::sleep_for(std::chrono::milliseconds(300));
```
Why are these wait times *so* long?  Especially if this scan application is run multiple times this would take *forever*.

lpetre-ulb replied:
These are the wait times from the original `gbt.py` tool : https://github.com/cms-gem-daq-project/reg_utils/blob/7b9cf050c63dae5f3a9b5803389130cff2f039fa/python/reg_interface/arm/gbt_utils.py#L69-L72
I don't know if they can reliably reduced.

lpetre-ulb replied:
Ok, I'll send him an email. In the meantime, I'll leae the wait times untouched.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+std::vector<uint32_t> scanGBTPhaseLocal(localArgs *la, const uint32_t ohN, const uint8_t phaseMin, const uint8_t phaseMax, const uint8_t phaseStep){
+    LOGGER->log_message(LogManager::INFO, stdsprintf("Scannng the phases for OH #%u.", ohN));
+
+    std::vector<uint32_t> results;
```
In the function documentation you have not described the contents, nor the order, of this container.  Use the `\returns` field in `doxygen`.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    // phases config keys
+    const uint8_t phaseMin = request->get_word("phaseMin");
+    const uint8_t phaseMax = request->get_word("phaseMax");
+    const uint8_t phaseStep = request->get_word("phaseStep");
```
Should *also* check here that `phaseMax` makes sense.  Keep the check in both places since the local callable function can be called elsewhere.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    // Perform the scan
+    std::vector<uint32_t> results = scanGBTPhaseLocal(&la, ohN, phaseMin, phaseMax, phaseStep);
+
+    response->set_word_array("results", results);
```
Needs to be explained in function documentation.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    // phase keys
+    const uint8_t phaseMin = request->get_word("phaseMin");
+    const uint8_t phaseMax = request->get_word("phaseMax");
+    const uint8_t phaseStep = request->get_word("phaseStep");
```
check that `phaseMax` makes sense.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+    return;
+} //Enc scanGBTPhase
+
+void scanGBTPhaseMultiLink(const RPCMsg *request, RPCMsg *response){
```
Again was expecting this to take a parameter that would define the number of scans to be performed in *this* rpc call.

bdorney commented on src/gbt.cpp:
```diff
@@ -0,0 +1,381 @@
+        }
+    }    
+
+    response->set_word_array("results", results);
```
documentation needed

bdorney commented:
Status update on this PR?

lpetre-ulb commented:
Yes, I tested the modifications on the hardware this afternoon. The methods still work as expected. As I better understand what you expect, I preferred to start a new clean PR which addresses the points you raised : https://github.com/cms-gem-daq-project/ctp7_modules/pull/75.
